### PR TITLE
CosmosDB RP API 2024-11-15 Version

### PR DESCRIPTION
**Pull request by pjohari-ms**
# ARM (Control Plane) API Specification Update Pull Request 

> [!TIP]
> Overwhelmed by all this guidance? See the `Getting help` section at the bottom of this PR description.

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

spec_pr_review_workflow_diagram

## Purpose of this PR

What's the purpose of this PR? Check the specific option that applies. This is **mandatory**!

  - [ ] New resource provider.
  - [x] New API version for an existing resource provider. (If API spec is not defined in TypeSpec, the PR should have been created in adherence to [OpenAPI specs PR creation guidance](https://aka.ms/azsdkdocs/createopenapispec)).
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix OpenAPI spec quality issues in S360.
  - [ ] Convert existing [OpenAPI spec to TypeSpec spec](https://aka.ms/typespec/conversion) (do not combine this with implementing changes for a new API version).
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [x] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://aka.ms/azurerpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to the diagram Step 2, "ARM API changes review", for this PR.

## Additional information

<details>
<summary> Viewing API changes</summary>

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

</details>
<details>
<summary>Suppressing failures</summary>

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[suppressions guide](https://aka.ms/azsdk/pr-suppressions) to get approval.

</details>

## Getting help

- First, please carefully read through this PR description, from top to bottom. Please fill out the `Purpose of this PR` and `Due diligence checklist`.
- If you don't have permissions to remove or add labels to the PR, request `write access` per [aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories](https://aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories)
- To understand what you must do next to merge this PR, see the `Next Steps to Merge` comment. It will appear within few minutes of submitting this PR and will continue to be up-to-date with current PR state.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- For help with ARM review (PR workflow diagram Step 2), see https://aka.ms/azsdk/pr-arm-review.
- If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
  This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

### CODE REVIEW

**Comment by ramoka178** [comment]
>         "enablePerRegionPerPartitionAutoscale": {

Standard guidance is: replace boolean/switch properties with a more meaningful enum whenever possible. 

A boolean will forever have two valid values (true or false). A string enum type is always preferred. Also, properties should always provide better values just than True and False. For example two switches "isTypeA" and "isTypeB" should be replaced with one enum "type": [A, B, DefaultType]. Enums are always a more flexible and future proof option because they allow additional values to be added in the future in a non-breaking way, e.g. [Enabled, Disabled, Suspended, Deallocated]. 

Note: do NOT define a 'boolean enum' with two values 'True and False'. This might be easier to 'extend' in terms of types, but semantically its cofusing, and no better than a boolean. 

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2024-11-15/cosmos-db.json:7041 in 9bbe7c6. [](commit_id = 9bbe7c61b01b661f3dddcf2b0d42d07eb46e276c, deletion_comment = False)

**Comment by ramoka178** [comment]
>         "enablePerRegionPerPartitionAutoscale": {

Please use enum as suggested in other comment

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2024-11-15/cosmos-db.json:7203 in 9bbe7c6. [](commit_id = 9bbe7c61b01b661f3dddcf2b0d42d07eb46e276c, deletion_comment = False)

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2024-11-15/notebook.json [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+          "200": {
+            "description": "Successfully retrieved notebook workspace resources.",
+            "schema": {
+              "$ref": "#/definitions/NotebookWorkspaceListResult"
+            }
+          },
+          "default": {
+            "description": "Error response describing why the operation failed.",
+            "schema": {
+              "$ref": "../../stable/2024-11-15/cosmos-db.json#/definitions/ErrorResponse"
```
Comment (ramoka178) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2024-11-15/notebook.json:71]: >../../stable/2024-11-15/cosmos-db.json#/definitions/ErrorResponse [](http://example.com/codeflow?start=23&length=65)

Is this version of ErrorResponse going to continue for all new api versions ? Common-types v5 has an ErrorResponse that we are pushing for Azure to use. 

Could you check the feasibility of referring that ? Here and other places of this PR.

**Comment by ramoka178** [comment]
>         "enablePerRegionPerPartitionAutoscale": {

Please use enum as suggested in other comment

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2024-11-15/cosmos-db.json:7389 in 9bbe7c6. [](commit_id = 9bbe7c61b01b661f3dddcf2b0d42d07eb46e276c, deletion_comment = False)

**Comment by ramoka178** [comment]
Please fix Swagger PrettierCheck failure too.

**Comment by ramoka178** [comment]
@pjohari-ms Please fix  Swagger ModelValidation and LintDiff errors. Also, please address the open comments that came up in previous review iteration

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2024-11-15/notebook.json (hunk 1 above)
Comment (pjohari-ms) [reply specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2024-11-15/notebook.json]: Done.

**Comment by pjohari-ms** [comment]
> > ```
> >     "enablePerRegionPerPartitionAutoscale": {
> > ```
> 
> Standard guidance is: replace boolean/switch properties with a more meaningful enum whenever possible.
> 
> A boolean will forever have two valid values (true or false). A string enum type is always preferred. Also, properties should always provide better values just than True and False. For example two switches "isTypeA" and "isTypeB" should be replaced with one enum "type": [A, B, DefaultType]. Enums are always a more flexible and future proof option because they allow additional values to be added in the future in a non-breaking way, e.g. [Enabled, Disabled, Suspended, Deallocated].
> 
> Note: do NOT define a 'boolean enum' with two values 'True and False'. This might be easier to 'extend' in terms of types, but semantically its cofusing, and no better than a boolean.
> 
> Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2024-11-15/cosmos-db.json:7041 in [9bbe7c6](https://github.com/Azure/azure-rest-api-specs/commit/9bbe7c61b01b661f3dddcf2b0d42d07eb46e276c). [](commit_id = [9bbe7c6](https://github.com/Azure/azure-rest-api-specs/commit/9bbe7c61b01b661f3dddcf2b0d42d07eb46e276c), deletion_comment = False)

I agree with the standard guidance. This parameter will always be enable/disable, so keeping it as a boolean for now.

**Comment by ramoka178** [comment]
> > > ```
> > >     "enablePerRegionPerPartitionAutoscale": {
> > > ```
> > 
> > 
> > Standard guidance is: replace boolean/switch properties with a more meaningful enum whenever possible.
> > A boolean will forever have two valid values (true or false). A string enum type is always preferred. Also, properties should always provide better values just than True and False. For example two switches "isTypeA" and "isTypeB" should be replaced with one enum "type": [A, B, DefaultType]. Enums are always a more flexible and future proof option because they allow additional values to be added in the future in a non-breaking way, e.g. [Enabled, Disabled, Suspended, Deallocated].
> > Note: do NOT define a 'boolean enum' with two values 'True and False'. This might be easier to 'extend' in terms of types, but semantically its cofusing, and no better than a boolean.
> > Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2024-11-15/cosmos-db.json:7041 in [9bbe7c6](https://github.com/Azure/azure-rest-api-specs/commit/9bbe7c61b01b661f3dddcf2b0d42d07eb46e276c). [](commit_id = [9bbe7c6](https://github.com/Azure/azure-rest-api-specs/commit/9bbe7c61b01b661f3dddcf2b0d42d07eb46e276c), deletion_comment = False)
> 
> I agree with the standard guidance. This parameter will always be enable/disable, so keeping it as a boolean for now.

Taking the words of another reviewer who gave a good explanation on this "This feedback and these linter checks aren't simply about the number of states/values supported by booleans vs. enums. It's about the design of the overall API itself. There are several reasons that booleans are essentially never a better choice than well-chosen enums.

For one, APIs tend to be updated over time. The approach of adding booleans to support new options/scenarios can cause APIs to collect booleans that reduce their discoverability, understandability, readability, and usability over time. I've seen cases where APIs start to look like bit vectors, and the code that uses them has to supply a list of true/false values. This makes it hard to understand what the call is actually doing (was that supposed to be operation(true, true, false, true, false, false) or operation(true, true, false, true, true, false)?).

At minimum, with an enum you can give better names to the values and provide individual descriptions of the meaning each value (even if there are only ever two). Often a well-designed enum can even replace multiple boolean properties and eliminate invalid/nonsense combinations. Doing this well requires thinking from the client perspective rather than the service perspective, although the service also often benefits from having an easier upgrade path: enums are conducive to adding scenarios by adding enum values over time, especially if they have been well thought out in previous versions. Enums seamlessly move from 2 to more than two states, whereas booleans force you to add more booleans to move past two states"

Since these are new properties, please check the suggestions again and address the issue.

**Comment by ramoka178** [comment]
Swagger ModelValidation  Avocado  errors are also failing. Please fix them.

**Comment by pjohari-ms** [comment]
@raosuhas I have fixed all issues - the avocado check is failing because of APIs not present in an old API version. Is it possible to add a permanent suppression for this?

**Review by razvanbadea-msft — APPROVED** [review]

**Pull request merged**